File: Daemons/leds.py
"""Tests that the LEDs are working, by writing to the appropriate GPIO pins"""

from periphery import GPIO  # pylint: disable=W0403

IN = "in"
OUT = "out"
HIGH = "high"
LOW = "low"

CCP_OK = GPIO(pin=13, direction=OUT)
IED_OK = GPIO(pin=12, direction=OUT)
FAULT = GPIO(pin=11, direction=OUT)
CCP_DATA_TX = GPIO(pin=25, direction=OUT)
CCP_DATA_RX = GPIO(pin=24, direction=OUT)
IED_DATA_TX = GPIO(pin=5, direction=OUT)
IED_DATA_RX = GPIO(pin=4, direction=OUT)


def test_ccp_ok(level):
    """Tests that the CCP OK LED can be turned on"""
    return _test(CCP_OK, level)

def test_ied_ok(level):
    """Tests that the IED OK LED can be turned on"""
    return _test(IED_OK, level)

def test_fault(level):
    """Tests that the Fault LED can be turned on"""
    return _test(FAULT, level)

def test_ccp_data_tx(level):
    """Tests that the CCP Data Tx (transmit) LED can be turned on"""
    return _test(CCP_DATA_TX, level)

def test_ccp_data_rx(level):
    """Tests that the CCP Data Rx (receive) LED can be turned on"""
    return _test(CCP_DATA_RX, level)

def test_ied_data_tx(level):
    """Tests that the IED Data Tx (transmit) LED can be turned on"""
    return _test(IED_DATA_TX, level)

def test_ied_data_rx(level):
    """Tests that the IED Data Rx (receive) LED can be turned on"""
    return _test(IED_DATA_RX, level)

def _test(pin, level):
    pin.write(level)
    if pin.read() is level:
        return True
    return False

File: PIM_Mini_Tests_WPF/LEDS/LEDS.cs
using PIM_Mini_Tests_WPF.Common;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PIM_Mini_Tests_WPF.LEDS
{
    public class LEDS : HardwareTest
    {
        public LEDS() : base("LEDs", new HardwareTest[] 
        {
            new TestPower(),
            new TestWatchdog(),
            new TestCcpOk(),
            new TestIedOk(),
            new TestFault(),
            new TestCcpDataTx(),
            new TestCcpDataRx(),
            new TestIedDataTx(),
            new TestIedDataRx()
        })
        {
        }


        /// <summary>
        /// Sets LEDs
        /// </summary>
        /// <param name="caller">The caller instance</param>
        /// <param name="level">true is on, false is off</param>
        /// <returns></returns>
        internal void Check(HardwareTest caller)
        {
            var result = this.SetLED(caller, true);
            if (result == DaemonResponse.Success)
            {
                bool userInput = caller.GetUserInput($"Is the {caller.Name} LED on?");
                caller.AssertEqual(userInput, true, $"The {caller.Name} LED could not be turned on.");
            }
            else
            {
                caller.AssertEqual(result, DaemonResponse.Success, $"Could not set the {caller.Name} LED"); // this fails the test
            }
            var resetResult = this.SetLED(caller, false);
            if (!caller.AssertEqual(resetResult.ToString(), DaemonResponse.Success.ToString(), $"The {caller.Name} LED could not be turned off")) return;
            caller.TestStatus = Status.Passed;
        }

        internal DaemonResponse SetLED(HardwareTest caller, bool level)
        {
            string ledLevel = level == true ? "on" : "off";
            return Controller.SendTcpMessage("LED_" + caller.Name.Remove(' ') + "_" + ledLevel);
        }

        public override void Test()
        {
            throw new NotImplementedException();
        }
    }
    internal enum TargetLEDs
    {
        CCP_OK,
        IED_OK,
        FAULT,
        CCP_DATA_TX,
        CCP_DATA_RX,
        IED_DATA_TX,
        IED_DATA_RX
    }
}
